ssl/statem/extensions_clnt.c

```diff
@@ -1254,7 +1254,30 @@ EXT_RETURN tls_construct_ctos_post_handshake_auth(SSL *s, WPACKET *pkt,
 #endif
 }
 
+#ifndef OPENSSL_NO_QUIC
+/* SAME AS tls_construct_stoc_quic_transport_params() */
+EXT_RETURN tls_construct_ctos_quic_transport_params(SSL *s, WPACKET *pkt,
+                                                    unsigned int context, X509 *x,
+                                                    size_t chainidx)
+{
+    if (s->ext.quic_transport_params == NULL
+        || s->ext.quic_transport_params_len == 0) {
+        return EXT_RETURN_NOT_SENT;
+    }
 
+    if (!WPACKET_put_bytes_u16(pkt, TLSEXT_TYPE_quic_transport_parameters)
+            || !WPACKET_start_sub_packet_u16(pkt)
+            || !WPACKET_sub_memcpy_u16(pkt, s->ext.quic_transport_params,
+                                       s->ext.quic_transport_params_len)
```

Comment:
Both `WPACKET_start_sub_packet_u16` and `WPACKET_sub_memcpy_u16` prefix their contents with the length. If we compare this to the BoringSSL equivalent (See https://github.com/google/boringssl/blob/ef0183c537536b19c2d7352a68cd3c2bbe5572c4/ssl/t1_lib.cc#L2655), we can see that the length is only prefixed once.

If we look at https://github.com/openssl/openssl/blob/0d03acea7aa45e94903fb12186ed6cc324eb1b03/include/internal/packet.h#L853, it calls `WPACKET_sub_memcpy__` which internally calls `WPACKET_start_sub_packet_len__` which `WPACKET_start_sub_packet_u16` does as well (See https://github.com/openssl/openssl/blob/0d03acea7aa45e94903fb12186ed6cc324eb1b03/include/internal/packet.h#L727) so I think you can fix the issue by removing the `WPACKET_start_sub_packet_u16` call and the `WPACKET_close` call.

Reply:
I looked at the decoder and you might be right that no changes are needed to the decoder. However, if no changes are needed to the decoder, I'm wondering how your transport parameter decoding currently works before this change. If no change to the decoder is needed, shouldn't you be getting double length prefix encoded transport parameters right now?

Reply:
I think the `PACKET_as_length_prefixed_2` call is unnecessary. If we look at the maxfragmentlen parser (https://github.com/openssl/openssl/blob/cd0fb43cbe2774220f2702b8289faec590a72d01/ssl/statem/extensions_srvr.c#L175), we see it doesn't do the `PACKET_as_length_prefixed_2` which means the ``PACKET_as_length_prefixed_2` call is parsing the encoded transport parameters length which should not happen in OpenSSL but in the QUIC implementation.